packages/grid/x-data-grid/src/colDef/gridDateColDef.ts

```diff
@@ -5,21 +5,45 @@ import { GridColTypeDef } from '../models/colDef/gridColDef';
 import { renderEditDateCell } from '../components/cell/GridEditDateCell';
 import { GridValueFormatterParams } from '../models/params/gridCellParams';
 
-export function gridDateFormatter({ value }: GridValueFormatterParams<Date | string>) {
-  if (value instanceof Date) {
-    return value.toLocaleDateString();
+function throwIfNotDateObject({
+  value,
+  columnType,
+  rowId,
+  field,
+}: {
+  value: any;
+  columnType: string;
+  rowId: any;
+  field: string;
+}) {
+  if (!(value instanceof Date)) {
+    throw new Error(
+      [
+        `MUI: \`${columnType}\` column type only accepts \`Date\` objects as values.`,
+        'Use `valueGetter` to transform the value into a `Date` object.',
+        `Row ID: ${rowId}, field: "${field}".`,
+      ].join('\n'),
+    );
   }
-  return value ?? '';
 }
 
-export function gridDateTimeFormatter({ value }: GridValueFormatterParams<Date | string>) {
-  if (value instanceof Date) {
-    return value.toLocaleString();
+export function gridDateFormatter({ value, field, id }: GridValueFormatterParams<Date>) {
+  if (!value) {
+    return '';
   }
-  return value ?? '';
+  throwIfNotDateObject({ value, columnType: 'date', rowId: id, field });
+  return value.toLocaleDateString();
 }
 
-export const GRID_DATE_COL_DEF: GridColTypeDef<Date | string, string> = {
+export function gridDateTimeFormatter({ value, field, id }: GridValueFormatterParams<Date>) {
+  if (!value) {
+    return '';
+  }
+  throwIfNotDateObject({ value, columnType: 'dateTime', rowId: id, field });
```

Comment:
Should we add a test case for checking this safety check is working as expected?

Reply:
I'll add it in a follow-up PR, didn't manage to add it quickly